### PR TITLE
Actually faster docker builds - Part 2

### PR DESCRIPTION
**Pull request by jonathangreen**
## Description

This PR updates our main docker build to use the new base image added in #1187. Since most of the logic has been moved into the base image build, it does a bit of refactoring to the main `Dockerfile`, so that all the logic is contained there, instead of being spread over a number of different scripts.

This also updates the build process for the base image slightly, so its easier to make and test changes to the base image in a PR. If the baseimage has been updated in a PR then `test-build.yml` will build the baseimage, and then use that base image for the container builds. If not, then we just use the prebuilt baseimage. 

This should get most of our docker builds to complete in a couple minutes, rather then the ~45 minute builds we currently have.

## Motivation and Context

I was frustrated with our slow docker build times, while trying to build images for nginx.

## How Has This Been Tested?

I've built the images and done some testing, spinning them up with `docker-compose` and doing some basic testing to make sure they are functional. I've also tested the CI part of this in my fork of the circulation repo where I can test pushing to `main` with reckless abandon.

## Checklist

- [X] I have updated the documentation accordingly.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by jonathangreen** [comment]
@io7m pinging you on this one, since you took a look at https://github.com/ThePalaceProject/circulation/pull/1187.

File: docker/services/nginx/nginx.conf
Comment (RishiDiwanTT) [inline docker/services/nginx/nginx.conf:1]: This is an entirely new configuration, why has it been introduced?

**Review by RishiDiwanTT — COMMENTED** [review]
Just have the one question about nginx.

File: docker/services/nginx/nginx.conf
Comment (jonathangreen) [reply docker/services/nginx/nginx.conf:1]: It's not really related to the speed improvements, but to the refactor of the docker file. 

The script we previously used to setup nginx used to echo a couple lines into the nginx config when we were setting it up the copied over the site config. 

I wanted to be able to just use a copy command to setup the whole nginx configuration, so I took the approach of copying over the configuration rather then echoing lines into it.

**Review by RishiDiwanTT — APPROVED** [review]

**Pull request merged**